geozero/src/wkt/wkt_reader.rs

```diff
@@ -0,0 +1,344 @@
+        read_wkt(&mut self.0.as_bytes(), processor)
+    }
+}
+
+/// Wkt Reader.
+pub struct WktReader<'a, R: Read>(pub &'a mut R);
+
+impl<'a, R: Read> GeozeroDatasource for WktReader<'a, R> {
+    fn process<P: FeatureProcessor>(&mut self, processor: &mut P) -> Result<()> {
+        read_wkt(&mut self.0, processor)
+    }
+}
+
+/// Read and process WKT geometry.
+pub fn read_wkt<R: Read, P: GeomProcessor>(reader: &mut R, processor: &mut P) -> Result<()> {
+    use std::str::FromStr;
+    // PERF: it would be good to avoid copying data into this string when we already
+    // have a string as input. Maybe the wkt crate needs a from_reader implementation.
+    let mut wkt_string = String::new();
+    reader.read_to_string(&mut wkt_string)?;
+    let wkt = wkt::Wkt::from_str(&wkt_string).map_err(|e| GeozeroError::Geometry(e.to_string()))?;
+    process_wkt_geom(&wkt.item, processor)
+}
+
+/// Process WKT geometry
+fn process_wkt_geom<P: GeomProcessor>(geometry: &Geometry<f64>, processor: &mut P) -> Result<()> {
+    process_wkt_geom_n(geometry, 0, processor)
+}
+
+fn process_wkt_geom_n<P: GeomProcessor>(
+    geometry: &Geometry<f64>,
+    idx: usize,
+    processor: &mut P,
+) -> Result<()> {
+    let multi_dim = processor.multi_dim();
+    match geometry {
+        Geometry::Point(g) => process_point(g, multi_dim, idx, processor)?,
+        Geometry::MultiPoint(g) => process_multi_point(g, idx, processor)?,
+        Geometry::LineString(g) => process_linestring(g, true, idx, processor)?,
+        Geometry::MultiLineString(g) => process_multilinestring(g, idx, processor)?,
+        Geometry::Polygon(g) => process_polygon(g, true, idx, processor)?,
+        Geometry::MultiPolygon(g) => process_multi_polygon(g, idx, processor)?,
+        Geometry::GeometryCollection(g) => process_geometry_collection(g, idx, processor)?,
+    }
+    Ok(())
+}
+
+fn process_geometry_collection<P: GeomProcessor>(
+    geometry_collection: &GeometryCollection<f64>,
+    idx: usize,
+    processor: &mut P,
+) -> Result<()> {
+    processor.geometrycollection_begin(geometry_collection.0.len(), idx)?;
+    for (idxg, geometry) in geometry_collection.0.iter().enumerate() {
+        process_wkt_geom_n(geometry, idxg, processor)?;
+    }
+    processor.geometrycollection_end(idx)
+}
+
+fn process_coord<P: GeomProcessor>(
+    coord: &Coord<f64>,
+    multi_dim: bool,
+    idx: usize,
+    processor: &mut P,
+) -> Result<()> {
+    if multi_dim {
+        processor.coordinate(coord.x, coord.y, coord.z, coord.m, None, None, idx)
+    } else {
+        processor.xy(coord.x, coord.y, idx)
+    }
+}
+
+fn process_point<P: GeomProcessor>(
+    point: &Point<f64>,
+    multi_dim: bool,
+    idx: usize,
+    processor: &mut P,
+) -> Result<()> {
+    processor.point_begin(idx)?;
+    if let Some(ref coord) = point.0 {
+        process_coord(coord, multi_dim, 0, processor)?;
+    } else {
+        return Err(GeozeroError::Geometry(
```

Comment:
AFAIK WKT is the only format with an "empty point".

I think erroring here is the safest thing to do here given that there is no support for it in other formats, so it's not really defined how they should handle it.

Another option would be to add
```
pub trait GeomProcessor {
...
    fn empty_coordinate() -> Result {
         Err(GeozeroError::Geometry("The WKT Point was empty, but Points cannot be empty in other formats".to_string()))
    }
}
```

And then only override it in the WktReader/WktWriter, but I'm not sure if it would be useful for anything. WDYT @pka?

Reply:
Paul Ramsey from PostGIS calls ignoring empty geometries strategic laziness (https://gis.stackexchange.com/a/106942). In the meantime PostGIS supports empty geometries and so does GEOS AFAIK. I would continue with strategic laziness and choose the cheapest solution. What happens, if we emit no coordinate and don't return an error?

Reply:
I tried being as lazy as possible with https://github.com/georust/geozero/pull/20/commits/dc6a63a15962970405623e02397cb0af6d2dd947

But I didn't like the results.
```
    #[test]
     fn empty_point() {
         let wkt = WktStr("POINT EMPTY");
         let actual = wkt.to_geo().unwrap();

         // This is weird - but it's not obvious to me what we should do in this case.
         let expected: geo_types::Geometry<f64> = point!(x: 0.0, y: 0.0).into();

         assert_eq!(expected, actual);
     }
```

We can't just "skip" an empty point - since the return type is not optional. It strikes me as pretty bad to return (0, 0). 

I've doubled back and reimplemented a "return error" approach when trying to process an empty point into anything other than that WKTWriter.

WDYT?

Reply:
What about representing empty points as ` geo_types::Geometry<f64> = point!(x: f64::NaN, y: f64::NaN).into();`  ?

Reply:
`geo` algorithms assume the absence of NaN values in order to produce correct results, and they _don't_ check for it (for performance reasons), so this would be a source of catastrophic downstream errors (what's the Euclidean distance between `Point(1.0)` and `Point(NaN, NaN)`) and / or a catastrophic perf regression (e.g. checking 10e7 points for NaN values)